packages/coverage-c8/src/provider.ts

```diff
@@ -6,6 +6,7 @@ import c from 'picocolors'
 import { provider } from 'std-env'
 import type { RawSourceMap } from 'vite-node'
 import { coverageConfigDefaults } from 'vitest/config'
+import { resolveReporters, updateThresholds } from '@vitest/utils/coverage'
```

Comment:
Maybe instead of using `utils` we can define a base coverage provider that others can extend from? So custom ones will also have access to it.

Reply:
Sure, we can create a `packages/coverage-base` that can be used for code sharing between other coverage packages. But the `coverage-base` cannot be a `CoverageProvider` implementation. It would just expose/export these `resolveReporters` and `updateThresholds` methods that istanbul and c8 providers can then call in their methods.
So basically it would be just a utility package for coverage providers. But we can make it a class that other providers can then `extends` from.

Something like

```ts

export class BaseCoverageProvider { // Does not implement CoverageProvider
  updateThresholds() {}
  resolveReporters() {}
}

export class C8CoverageProvider extends BaseCoverageProvider implements CoverageProvider {
  initialize(ctx: Vitest) {
    this.options = resolveC8Options(ctx.config.coverage, ctx.config.root, this.resolveReporters())
  }
  reportCoverage() {
    // ... c8 related magic ...
    this.updateThresholds();
  }
}
```

Reply:
> Sure, we can create a `packages/coverage-base`

No, what I mean is to add this to `vitest` core and export it in coverages under `vitest/coverage`. We already expose coverage interface and config.

Reply:
Adding this to `vitest` core package is possible. But it should be noted that every time something new is added there and the coverage providers rely on that, they need to bump the peer dependency version, and that is a breaking change.

Reply:
Sure

Reply:
The [d6bfeef](https://github.com/vitest-dev/vitest/pull/2886/commits/d6bfeefe016ad9093dc4a9d5638475caa035247c) adds `vitest/coverage` - does this like something you meant?

The build of coverage packages now fails for some type declaration error. I'm not sure what is causing this. Any pointers are appreciated. 🙌 

<details>
  <summary>Build error</summary>

```
packages/coverage-c8 build$ rimraf dist && rollup -c
│ 
│ src/index.ts → dist...
│ created dist in 68ms
│ 
│ src/index.ts → dist...
│ (!) Unresolved dependencies
│ https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
│ node:worker_threads (imported by ../vitest/dist/coverage.d.ts, ../vitest/dist/types-c6ea4de7.d.ts)
│ node:fs (imported by ../vitest/dist/coverage.d.ts, ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/
│ node:http (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node/
│ node:stream (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/nod
│ node:events (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/nod
│ node:net (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node/i
│ node:tls (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node/i
│ node:https (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node
│ node:url (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node/i
│ node:zlib (imported by ../../node_modules/.pnpm/vite@4.0.0_@types+node@18.7.13/node_modules/vite/dist/node/
│ [!] (plugin dts) Error: namespace child (hoisting) not supported yet
│   101 |
│   102 | declare module 'source-map-js/lib/source-map-generator' {
│ > 103 |     import { SourceMapGenerator } from 'source-map-js'
│       |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
│   104 |     export { SourceMapGenerator as SourceMapGenerator }
│   105 | }
│   106 |
│ ../../node_modules/.pnpm/source-map-js@1.0.2/node_modules/source-map-js/source-map.d.ts
│ Error: namespace child (hoisting) not supported yet
│   101 |
│   102 | declare module 'source-map-js/lib/source-map-generator' {
│ > 103 |     import { SourceMapGenerator } from 'source-map-js'
│       |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
│   104 |     export { SourceMapGenerator as SourceMapGenerator }
│   105 | }
│   106 |
│     at DeclarationScope.convertNamespace (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_
│     at Transformer.convertNamespaceDeclaration (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@
│     at Transformer.convertStatement (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob
│     at new Transformer (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob3xud5wukob6ph
│     at convert (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob3xud5wukob6phfmz2mbyy
│     at Object.transform (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob3xud5wukob6p
│     at transformFile (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob3xud5wukob6phfm
│     at Object.transform (/Users/ari/Git/vitest/node_modules/.pnpm/rollup-plugin-dts@4.2.3_ntuob3xud5wukob6p
│     at /Users/ari/Git/vitest/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/shared/rollup.js:228
└─ Failed in 11.8s at /Users/ari/Git/vitest/packages/coverage-c8
```

</details>